In [1]:
install.packages('nycflights13')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [2]:
options(repr.plot.width=8, repr.plot.height=5)
library(tidyverse)
library(nycflights13)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.3     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


# Lecture 5: Tidy data, missing data

<div style="border: 1px double black; padding: 10px; margin: 10px">

**After today's lecture you will:**
* Gain more experience creating and working with tidy data
* Learn about how R handles missing data.
    
This lecture note corresponds to Chapters 6 and 20 of your book.
</div>


    




## A usage of `pivot_wider()`
Let's revisit the `gapminder` dataset that we first saw last lecture:

In [3]:
install.packages('gapminder')

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [4]:
library(gapminder)
print(gapminder)

# A tibble: 1,704 × 6
   country     continent  year lifeExp      pop gdpPercap
   <fct>       <fct>     <int>   <dbl>    <int>     <dbl>
 1 Afghanistan Asia       1952    28.8  8425333      779.
 2 Afghanistan Asia       1957    30.3  9240934      821.
 3 Afghanistan Asia       1962    32.0 10267083      853.
 4 Afghanistan Asia       1967    34.0 11537966      836.
 5 Afghanistan Asia       1972    36.1 13079460      740.
 6 Afghanistan Asia       1977    38.4 14880372      786.
 7 Afghanistan Asia       1982    39.9 12881816      978.
 8 Afghanistan Asia       1987    40.8 13867957      852.
 9 Afghanistan Asia       1992    41.7 16317921      649.
10 Afghanistan Asia       1997    41.8 22227415      635.
# ℹ 1,694 more rows


## Missing data in R
The `gapminder` appears very nice, it only contains data sets with no missing observations. But! The raw data looks like this:

In [5]:
gapminder_unfiltered  %>% print

# A tibble: 3,313 × 6
   country     continent  year lifeExp      pop gdpPercap
   <fct>       <fct>     <int>   <dbl>    <int>     <dbl>
 1 Afghanistan Asia       1952    28.8  8425333      779.
 2 Afghanistan Asia       1957    30.3  9240934      821.
 3 Afghanistan Asia       1962    32.0 10267083      853.
 4 Afghanistan Asia       1967    34.0 11537966      836.
 5 Afghanistan Asia       1972    36.1 13079460      740.
 6 Afghanistan Asia       1977    38.4 14880372      786.
 7 Afghanistan Asia       1982    39.9 12881816      978.
 8 Afghanistan Asia       1987    40.8 13867957      852.
 9 Afghanistan Asia       1992    41.7 16317921      649.
10 Afghanistan Asia       1997    41.8 22227415      635.
# ℹ 3,303 more rows


### What happens when we reshape the "unfiltered" data?

In [6]:
# pivot unfiltered data wider
gapminder_unfiltered %>% pivot_wider(id_cols = country, names_from = year, values_from = gdpPercap)

country                1952       1957       1962       1967      
1   Afghanistan              779.4453   820.8530   853.1007   836.1971
2   Albania                 1601.0561  1942.2842  2312.8890  2760.1969
3   Algeria                 2449.0082  3013.9760  2550.8169  3246.9918
4   Angola                  3520.6103  3827.9405  4269.2767  5522.7764
5   Argentina               5911.3151  6856.8562  7133.1660  8052.9530
6   Armenia                        NA         NA         NA         NA
7   Aruba                          NA         NA         NA         NA
8   Australia              10039.5956 10949.6496 12217.2269 14526.1246
9   Austria                 6137.0765  8842.5980 10750.7211 12834.6024
10  Azerbaijan                     NA         NA         NA         NA
11  Bahamas                        NA         NA 15117.5150 19318.1061
12  Bahrain                 9867.0848 11635.7995 12753.2751 14804.6727
13  Bangladesh               684.2442   661.6375   686.3416   721.1861
14  Barbados                       NA         NA  6204.4895  7788.5441
15  Belarus                        NA         NA         NA         NA
16  Belgium                 8343.1051  9714.9606 10991.2068 13149.0412
17  Belize                         NA         NA  1916.3196  2099.1585
18  Benin                   1062.7522   959.6011   949.4991  1035.8314
19  Bhutan                         NA         NA         NA         NA
20  Bolivia                 2677.3263  2127.6863  2180.9725  2586.8861
21  Bosnia and Herzegovina   973.5332  1353.9892  1709.6837  2172.3524
22  Botswana                 851.2411   918.2325   983.6540  1214.7093
23  Brazil                  2108.9444  2487.3660  3336.5858  3429.8644
24  Brunei                         NA         NA         NA         NA
25  Bulgaria                2444.2866  3008.6707  4254.3378  5577.0028
26  Burkina Faso             543.2552   617.1835   722.5120   794.8266
27  Burundi                  339.2965   379.5646   355.2032   412.9775
28  Cambodia                 368.4693   434.0383   496.9136   523.4323
29  Cameroon                1172.6677  1313.0481  1399.6074  1508.4531
30  Canada                 11367.1611 12489.9501 13462.4855 16076.5880
⋮   ⋮                      ⋮          ⋮          ⋮          ⋮         
158 Suriname                       NA         NA         NA         NA
159 Swaziland               1148.3766  1244.7084  1856.1821  2613.1017
160 Sweden                  8527.8447  9911.8782 12329.4419 15258.2970
161 Switzerland            14734.2327 17909.4897 20431.0927 22966.1443
162 Syria                   1643.4854  2117.2349  2193.0371  1881.9236
163 Taiwan                  1206.9479  1507.8613  1822.8790  2643.8587
164 Tajikistan                     NA         NA         NA         NA
165 Tanzania                 716.6501   698.5356   722.0038   848.2187
166 Thailand                 757.7974   793.5774  1002.1992  1295.4607
167 Timor-Leste                    NA         NA         NA         NA
168 Togo                     859.8087   925.9083  1067.5348  1477.5968
169 Tonga                          NA         NA         NA         NA
170 Trinidad and Tobago     3023.2719  4100.3934  4997.5240  5621.3685
171 Tunisia                 1468.4756  1395.2325  1660.3032  1932.3602
172 Turkey                  1969.1010  2218.7543  2322.8699  2826.3564
173 Turkmenistan                   NA         NA         NA         NA
174 Uganda                   734.7535   774.3711   767.2717   908.9185
175 Ukraine                        NA         NA         NA         NA
176 United Arab Emirates           NA         NA         NA         NA
177 United Kingdom          9979.5085 11283.1779 12477.1771 14142.8509
178 United States          13990.4821 14847.1271 16173.1459 19530.3656
179 Uruguay                 5716.7667  6150.7730  5603.3577  5444.6196
180 Uzbekistan                     NA         NA         NA         NA
181 Vanuatu                        NA         NA         NA         NA
182 Venezuela               

In [ ]:
# pivot unfiltered data wider
gapminder_unfiltered %>% pivot_wider(id_cols = country, names_from = year, values_from = gdpPercap)

# A tibble: 187 × 59
   country `1952` `1957` `1962` `1967` `1972` `1977` `1982` `1987` `1992` `1997`
   <fct>    <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>  <dbl>
 1 Afghan…   779.   821.   853.   836.   740.   786.   978.   852.   649.   635.
 2 Albania  1601.  1942.  2313.  2760.  3313.  3533.  3631.  3739.  2497.  3193.
 3 Algeria  2449.  3014.  2551.  3247.  4183.  4910.  5745.  5681.  5023.  4797.
 4 Angola   3521.  3828.  4269.  5523.  5473.  3009.  2757.  2430.  2628.  2277.
 5 Argent…  5911.  6857.  7133.  8053.  9443. 10079.  8998.  9140.  9308. 10967.
 6 Armenia    NA     NA     NA     NA     NA     NA     NA     NA   1443.  1791.
 7 Aruba      NA     NA     NA     NA   4940.  7390. 10875. 17674. 25121. 26484.
 8 Austra… 10040. 10950. 12217. 14526. 16789. 18334. 19477. 21889. 23425. 26998.
 9 Austria  6137.  8843. 10751. 12835. 16662. 19749. 21597. 23688. 27042. 29096.
10 Azerba…    NA     NA     NA     NA     NA     NA     NA     NA   3456.  1943.
# ℹ 177

You can see that there are many missing observations in the unfiltered data. In real life, you will mostly get unfiltered data -- how should we handle missing data?

## Missing Values
Missing values can be:

* **Explicit** (marked as `NA` in our data); or
* **Implicit** (not present in the data).

In [11]:
df <- tribble(
  ~person,           ~treatment, ~response,
  "Derrick Whitmore", 1,         7,
  NA,                 2,         10,
  NA,                 3,         NA,
  "Katherine Burke",  1,         4,
  NA,  1,         NA
)

In [12]:
df

person           treatment response
1 Derrick Whitmore 1          7      
2 NA               2         10      
3 NA               3         NA      
4 Katherine Burke  1          4      
5 NA               1         NA

The missing values are **explicit** in this table: each missing value is indicated by `NA` in the table.

You can fill in these missing values with `tidyr::fill()`. It works like `select()`, taking a set of columns, and fills them in with the last non-missing value.

In [13]:
# fill df for person and response columns
fill(df, person, response)

person           treatment response
1 Derrick Whitmore 1          7      
2 Derrick Whitmore 2         10      
3 Derrick Whitmore 3         10      
4 Katherine Burke  1          4      
5 Katherine Burke  1          4

In [10]:
# fill in each missing value in gapminder_unfiltered carrying forward
fill(gapminder_unfiltered, gdpPercap) %>% filter(is.na(gdpPercap))

country continent year lifeExp pop gdpPercap

## Encoded missing data

Sometimes, data contain a concrete value that actually represents a missing value. You see this often when dealing with data that is imported from a format that does not have a way to represent missing values, such as text or CSV. So it must instead use some special value like 99 or -999.

To correct for this type of missing value, we can use a function called `na_if(x, n)`. This takes a vector `x` and replaces any occurence of `n` with `NA`:

In [15]:
x = c(-99, 1, 3, -99, 2)
print(x)
na_if(x, -99)

[1] -99   1   3 -99   2


[1] NA  1  3 NA  2

## Implicit missing values
A second type of missing data occurs when there are simply no observations in the dataset for a particular combination of columns. For example:

In [16]:
stocks <- tibble(
  year  = c(2020, 2020, 2020, 2020, 2021, 2021, 2021),
  qtr   = c(   1,    2,    3,    4,    2,    3,    4),
  price = c(1.88, 0.59, 0.35,   NA, 0.92, 0.17, 2.66)
)

In [17]:
print(stocks)

# A tibble: 7 × 3
   year   qtr price
  <dbl> <dbl> <dbl>
1  2020     1  1.88
2  2020     2  0.59
3  2020     3  0.35
4  2020     4 NA   
5  2021     2  0.92
6  2021     3  0.17
7  2021     4  2.66


This dataset has two missing observations:
- The price in the 2020q4 is explicitly missing. (It has an `NA`.)
- The price in 2021q1 is implicitly missing: it does not appear in the dataset.

> An explicit missing value is the presence of an absence.
>
> An implicit missing value is the absence of a presence.


How can we handle implicit missing values? As we have already seen, one option is to use `pivot_wider()`:

In [18]:
# using pivot_wider() on stocks converts implicit missing values to explicit
pivot_wider(stocks, names_from = qtr, values_from = price) %>%
  print

# A tibble: 2 × 5
   year   `1`   `2`   `3`   `4`
  <dbl> <dbl> <dbl> <dbl> <dbl>
1  2020  1.88  0.59  0.35 NA   
2  2021 NA     0.92  0.17  2.66


Alternatively, we can use the `complete()` function, which creates entries for all possible combinations of a set of columns:

In [19]:
# use complete to fill in missing values for stocks
complete(stocks, year, qtr)

year qtr price
1 2020 1   1.88 
2 2020 2   0.59 
3 2020 3   0.35 
4 2020 4     NA 
5 2021 1     NA 
6 2021 2   0.92 
7 2021 3   0.17 
8 2021 4   2.66

### Fixed value replacement

Sometimes missing values represent some fixed and known value, most commonly 0. You can use dplyr::coalesce() to replace them:

In [20]:
x <- c(1, 4, 5, 7, NA)
coalesce(x, 100)
#> [1] 1 4 5 7 0

[1]   1   4   5   7 100

### Titanic analysis

On April 15, 1912, the great "unsinkable" RMS Titanic ship sank to the bottom of the Atlantic.

Unfortunately, there weren’t enough lifeboats for everyone on board, resulting in more than 1500 deaths out of 2224 passengers and crew. Refer:https://en.wikipedia.org/wiki/Titanic

We have partial list of passengers here. Let us do some analysis

In [22]:
titanic = read.csv('https://storage.googleapis.com/mbcc/titanic.csv')
titanic %>% head %>% print

  survived pclass    sex age sibsp parch    fare embarked class   who
1        0      3   male  22     1     0  7.2500        S Third   man
2        1      1 female  38     1     0 71.2833        C First woman
3        1      3 female  26     0     0  7.9250        S Third woman
4        1      1 female  35     1     0 53.1000        S First woman
5        0      3   male  35     0     0  8.0500        S Third   man
6        0      3   male  NA     0     0  8.4583        Q Third   man
  adult_male deck embark_town alive alone
1       True      Southampton    no False
2      False    C   Cherbourg   yes False
3      False      Southampton   yes  True
4      False    C Southampton   yes False
5       True      Southampton    no  True
6       True       Queenstown    no  True


## &#129300; Quiz

What is the average age of the passengers (ignoring the decimal part)?

<ol style="list-style-type: upper-alpha;">
    <li>28</li>
    <li>30</li>
    <li>27</li>
    <li>29</li>
    <li>31</li>
</ol>


In [23]:
## solution
titanic %>%
  filter(!is.na(age)) %>%
  summarize(mean_age = mean(age))

mean_age
1 29.69912

In [24]:
summary(titanic$age)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
   0.42   20.12   28.00   29.70   38.00   80.00     177 

In [26]:
mean(titanic$age, na.rm = T)

[1] 29.69912

In [28]:
titanic$age %>% mean(na.rm = T)

[1] 29.69912

In [29]:
summary(titanic)

    survived          pclass          sex                 age       
 Min.   :0.0000   Min.   :1.000   Length:891         Min.   : 0.42  
 1st Qu.:0.0000   1st Qu.:2.000   Class :character   1st Qu.:20.12  
 Median :0.0000   Median :3.000   Mode  :character   Median :28.00  
 Mean   :0.3838   Mean   :2.309                      Mean   :29.70  
 3rd Qu.:1.0000   3rd Qu.:3.000                      3rd Qu.:38.00  
 Max.   :1.0000   Max.   :3.000                      Max.   :80.00  
                                                     NA's   :177    
     sibsp           parch             fare          embarked        
 Min.   :0.000   Min.   :0.0000   Min.   :  0.00   Length:891        
 1st Qu.:0.000   1st Qu.:0.0000   1st Qu.:  7.91   Class :character  
 Median :0.000   Median :0.0000   Median : 14.45   Mode  :character  
 Mean   :0.523   Mean   :0.3816   Mean   : 32.20                     
 3rd Qu.:1.000   3rd Qu.:0.0000   3rd Qu.: 31.00                     
 Max.   :8.000   Max.   :6.0

## &#129300; Quiz

How many na values are there in the age column?

<ol style="list-style-type: upper-alpha;">
    <li>177</li>
    <li>178</li>
    <li>714</li>
    <li>There are no NA values</li>
</ol>


#### multiple ways of finding the answer

In [30]:
sum(is.na(titanic$age))

[1] 177

In [33]:
titanic %>% count(is.na(age))

is.na(age) n  
1 FALSE      714
2  TRUE      177

In [34]:
titanic %>% filter(is.na(age)) %>% nrow

[1] 177

Replace the NA values in the `age` column with average age

In [37]:
## solution

titanic %>%
  mutate(age = coalesce(age, mean(age, na.rm = T))) %>%
  count(is.na(age))


is.na(age) n  
1 FALSE      891